concurrency/pool.go

```diff
@@ -0,0 +1,85 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+// WorkerPool ensures that for any number of concurrent ForEachNotInFlight calls each with any number of tokens only up to numWorkers f
+// invocations are executing concurrently. See the docs of ForEachNotInFlight for the uniqueness semantics of tokens.
+type WorkerPool struct {
+	inflightTokensMx sync.Mutex
+	inflightTokens   map[string]struct{}
+	inflightCalls    sync.WaitGroup
+	semaphore        chan struct{}
+}
+
+func NewWorkerPool(numWorkers int) *WorkerPool {
+	return &WorkerPool{
+		inflightTokensMx: sync.Mutex{},
+		inflightTokens:   make(map[string]struct{}),
+		inflightCalls:    sync.WaitGroup{},
+		semaphore:        make(chan struct{}, numWorkers),
+	}
+}
+
+// Wait returns when there are no in-flight calls to ForEachNotInFlight.
+func (w *WorkerPool) Wait() {
+	w.inflightCalls.Wait()
+}
+
+// ForEachNotInFlight invokes f for every token in tokens that is not in-flight (not still being executed) in a different
+// concurrent call to ForEachNotInFlight. ForEachNotInFlight returns when invocations to f for all such tokens have
+// returned. Upon context cancellation ForEachNotInFlight stops making new invocations of f for tokens and waits for all
+// already started invocations of f to return. ForEachNotInFlight returns the combined errors from all f invocations.
+func (w *WorkerPool) ForEachNotInFlight(ctx context.Context, tokens []string, f func(context.Context, string) error) error {
+	w.inflightCalls.Add(1)
+	defer w.inflightCalls.Done()
+
+	notInflightTokens := make([]string, 0, len(tokens))
+	w.inflightTokensMx.Lock()
+	for _, token := range tokens {
+		if _, ok := w.inflightTokens[token]; ok {
+			continue
+		}
+		notInflightTokens = append(notInflightTokens, token)
+		w.inflightTokens[token] = struct{}{}
+	}
+	w.inflightTokensMx.Unlock()
+
+	var (
+		errs    multierror.MultiError
+		errsMx  sync.Mutex
+		workers sync.WaitGroup
+	)
+
+	for _, token := range notInflightTokens {
+		select {
+		case <-ctx.Done():
+			workers.Wait()
+			return errs.Err()
+		case w.semaphore <- struct{}{}:
+		}
+
+		workers.Add(1)
+		go func(token string) {
```

Comment:
The name `WorkerPool` implies (to me) that it creates a pool of worker go routines which then execute the given functions, so the overhead of creating new go routines can be avoided. I'm surprised to see that this creates a new go function, so it's more like a way to concurrently execute functions while applying certain limitations such as max concurrency and a guarantee that each token is only executed once, right? 
Naming is hard...  but how about something like `LimitedConcurrencyExecutor` instead?

Reply:
I see what you mean. Naming is indeed hard.

> guarantee that each token is only executed once, right?

not exactly. it guarantees that each token will not be executed concurrently with itself ("with itself" by another worker).

`LimitedConcurrencyOnceExecutor` makes me think of `sync.Once`, which is a wrong association. [It's more like `x/sync.SingleFlight`](https://pkg.go.dev/golang.org/x/sync/singleflight). Would `LimitedConcurrencySingleFlightExecutor` be better?

or just `ConcurrentSingleFlight`. And rename the method to `Do(ctx, keys, f)`? I prefer this slightly

Reply:
> or just ConcurrentSingleFlight. 

Isn't the normal `sync.SingleFlight` also concurrent, with an unlimited concurrency as long as the keys are unique? So the main difference to `sync.SingleFlight` would be that here the concurrency is limited, right? How about `LimitedConcurrencySingleFlight`?

Reply:
`LimitedConcurrencySingleFlight` sounds good to me

Reply:
renamed in 053e70cbaaa8260a98beb8216409f0f819f4277f